Add private key password option to ssl adapters

jatalahd opened the pull request:
- It is now possible to use password protected private keys in both builtin and openssl ssl-adapters
- Added also positive and negative unit test cases
- With reference to https://github.com/cherrypy/cherrypy/issues/1583

❓ **What kind of change does this PR introduce?**

* [ ] 🐞 bug fix
* [X] 🐣 feature
* [ ] 📋 docs update
* [ ] 📋 tests/coverage improvement
* [ ] 📋 refactoring
* [ ] 💥 other

📋 **What is the related issue number (starting with `#`)**

https://github.com/cherrypy/cherrypy/issues/1583

❓ **What is the current behavior?** (You can also link to an open issue here)
With the current functionality it is only possible to use ssl adapters with private keys without password protection


❓ **What is the new behavior (if this is a feature change)?**
With this change, there is a new option to give the ssl adapter a "private_key_password" argument, which can be in either string or bytestring format.


📋 **Other information**:
Added also unit tests to test the new functionality


📋 **Contribution checklist:**


* [X] I wrote descriptive pull request text above
* [X] I think the code is well written
* [X] I wrote [good commit messages]
* [x] I have [squashed related commits together][related squash] after
      the changes have been approved
* [X] Unit tests for the changes exist
* [x] Integration tests for the changes exist (if applicable)
* [X] I used the same coding conventions as the rest of the project
* [X] The new code doesn't generate linter offenses
* [x] Documentation reflects the changes
* [X] The PR relates to *only* one subject with a clear title
      and description in grammatically correct, complete sentences

[good commit messages]: http://chris.beams.io/posts/git-commit/
[making a lovely PR]: https://mtlynch.io/code-review-love/
[related squash]:
https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cherrypy/cheroot/752)
<!-- Reviewable:end -->

jatalahd commented:
Fixed linter issues with second commit, but now it seems that I need to do the git squash... Anyway, it would be important for us to get this improvement into some upcoming cheroot release, which we can then reference in our pipfile so that we can take these changes into use.

webknjaz commented:
@jatalahd thanks for the contrib! You don't *have to* squash commits unconditionally. Only if the commits are non-atomic. If they aren't, it's usually a good idea to combine them to keep Git history clean.

One thing that's definitely missing is a change note. Read https://cheroot.cherrypy.dev/en/latest/contributing/guidelines/#adding-change-notes-with-your-prs and follow the guidelines. Do you best and if it needs editing, I'll tell you.

I can't give you a timeline right now. I'm a bit unhappy about the state of the TLS adapters in general and wanted to redesign them eventually. And so I'm a bit hesitant on what changes would be acceptable in the public API. I'll need to think about it first.

I'll leave a few notes in the diff but that'll be an incomplete review.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -726,3 +786,99 @@ def test_http_over_https_error(
+    httpserver.ssl_adapter = tls_adapter
+
+    # expecting SSLError when preparing server
+    with pytest.raises(OpenSSL.SSL.Error):
```
Always use the `match=` argument with `pytest.raises()`. This will also make the code comment unnecessary.

jatalahd replied:
Added `match=` argument, but the textual error messages are quite elaborous, so a partial regex match is now used.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -726,3 +786,99 @@ def test_http_over_https_error(
+    key_password,
+):
+    """Check that server init fails when wrong private key password given."""
+    key_file, cert_file = create_private_key_with_password
```
It's uncommon to use verbs in fixture names that don't return callables.

jatalahd replied:
The problematic fixture is now removed.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -118,6 +123,61 @@ def make_tls_http_server(bind_addr, ssl_adapter, request):
     return httpserver
 
 
+def create_self_signed_certificate(key_object):
```
Reworked the test fixtures so that removed the old methods and added one new that gets the private key from trustme and rewrites that with password protection using cryptography module. The rewritten file is removed at teardown (after yield) using `Path.unlink()` . Much cleaner than the previous, but not perfect. I cannot do better than this, tho.

jatalahd commented:
@webknjaz ; Thanks for reviewing the code. I fixed all problems you mentioned (as well as I could). Added also change note and had to do the git squash to clean up the git history. Due to that, I had to force push and therefore your review comments disappeared from the "files changed" -tab. However I answered all those comments in this conversation-tab. Please let me know if there is something that still needs fixing.

webknjaz commented on cheroot/ssl/pyopenssl.py:
```diff
@@ -328,13 +333,28 @@ def wrap(self, sock):
         # See https://code.activestate.com/recipes/442473/
         c = SSL.Context(SSL.SSLv23_METHOD)
+        if self.private_key_password is not None:
+            c.set_passwd_cb(self._password_callback, self.private_key_password)
```
https://www.pyopenssl.org/en/stable/api/ssl.html#OpenSSL.SSL.Context.set_passwd_cb says
> Set the passphrase callback. This function will be called when a private key with a passphrase is loaded.

This probably means that it can be called unconditionally and pyOpenSSL will ignore it when the private key is non-encrypted, right?
```suggestion
        c.set_passwd_cb(self._password_callback, self.private_key_password)
```

jatalahd replied:
Yes, it seems so. Modified the code like yo suggested and I added unit test for this use case.

jatalahd commented on cheroot/ssl/__init__.py:
```diff
@@ -20,12 +20,14 @@ def __init__(
         private_key,
         certificate_chain=None,
         ciphers=None,
+        private_key_password=None,
```
Done in latest push. Added the asterisk also to the .pyi file

jatalahd commented on cheroot/ssl/builtin.py:
```diff
@@ -112,15 +121,25 @@ def _loopback_for_cert(certificate, private_key, certificate_chain):
+    certificate,
+    private_key,
+    certificate_chain,
+    private_key_password=None,
```
Done in latest push.

jatalahd commented on cheroot/ssl/pyopenssl.py:
```diff
@@ -293,12 +293,16 @@ class pyOpenSSLAdapter(Adapter):
         private_key,
         certificate_chain=None,
         ciphers=None,
+        private_key_password=None,
```
Done in latest push.

jatalahd commented on cheroot/ssl/builtin.py:
```diff
@@ -208,6 +230,7 @@ def __init__(
         private_key,
         certificate_chain=None,
         ciphers=None,
+        private_key_password=None,
```
Done in latest push.

jatalahd commented on cheroot/ssl/pyopenssl.pyi:
```diff
@@ -31,9 +31,16 @@ class pyOpenSSLAdapter(Adapter):
+        self,
+        _password_max_length: int,
+        _verify_twice: bool,
+        password: bytes | str,
```
Done in latest push.

jatalahd commented on cheroot/ssl/pyopenssl.py:
```diff
@@ -328,13 +333,28 @@ def wrap(self, sock):
+        self,
+        _password_max_length,
+        _verify_twice,
+        password,
```
Done in latest push.

jatalahd commented on cheroot/ssl/pyopenssl.py:
```diff
@@ -328,13 +333,28 @@ def wrap(self, sock):
+            return b''
+        if not isinstance(password, bytes):
+            return password.encode('utf-8')
+        return password
```
Great suggestion, adopted this with minor style modifications (since the linter was giving errors). The latest push includes this.

webknjaz commented on cheroot/ssl/pyopenssl.py:
```diff
@@ -328,13 +335,38 @@ def wrap(self, sock):
+
+        if len(b_password) > password_max_length:
+            _warn(
+                f'User-provided password is {len(b_password)} bytes long and will '
```
I'd rather not have length computed twice. Also, complex expressions reduce readability. Could you move this to a var?

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -137,6 +148,14 @@ def make_tls_http_server(bind_addr, ssl_adapter, request):
+    """Provide random 10 character password for private key."""
+    return ''.join(
+        secrets.choice(string.ascii_letters + string.digits) for _ in range(10)
+    )
```
Let's just keep it simple
```suggestion
    """Return a password for private key."""
    return 'криївка'
```

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -751,3 +800,142 @@ def test_http_over_https_error(
+        'builtin',
+        'pyopenssl',
+    )
+    * 2,
```
Removed double run in the latest push.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -177,6 +196,36 @@ def tls_certificate_private_key_pem_path(tls_certificate):
+        keyfile_temp_path = Path(tf)
+    keyfile_temp_path.write_bytes(encrypted_key_as_bytes)
+
+    yield keyfile_temp_path
```
Moved under `with`. You commented on this already 4 months ago, and I managed to get this much cleaner since then. Now it uses the base certificate provided by trustme and only encrypts the output. After this change the yield part is similar to other certificate fixtures. I cannot do any better than this.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -751,3 +800,142 @@ def test_http_over_https_error(
+    httpserver.ssl_adapter = tls_adapter
+
+    with pytest.raises(OpenSSL.SSL.Error, match=r'.+bad decrypt.+'):
+        httpserver.prepare()
```
The builtin adapter prepares the server env with a call to already in `__init__` while pyopenssl uses the cert only when doing `bind`. Also it seems that the builtin adapter makes two calls to load the certs, so when starting the server manually, user needs to enter the password twice. So there is a clear difference how builtin and openssl adapters behave in this sense. I am not proficient enough to start making changes to unify the behaviour.

webknjaz replied:
Okay, that's fair. Marking this as resolved.

jatalahd commented on cheroot/test/test_ssl.py:
Done in latest push by parametering the last test case.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -751,3 +800,142 @@ def test_http_over_https_error(
+        'pyopenssl',
+    ),
+)
+def test_ssl_adapters_key_without_password(
```
Combined in the latest push.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +798,124 @@ def test_http_over_https_error(
+        gateway=Gateway,
+    )
+
+    key_file = request.getfixturevalue(key_file)
```
I don't understand the motivation here.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +798,124 @@ def test_http_over_https_error(
+    httpserver.ssl_adapter = tls_adapter
+
+    with expected_warn:
+        with pytest.raises(OpenSSL.SSL.Error, match=r'.+bad decrypt.+'):
```
I believe these can be collapsed into a single with-block.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +798,124 @@ def test_http_over_https_error(
+            ),
+        ),
+    ),
+    ids=('false_password', 'integer_password', 'too_long_password'),
```
```suggestion
    ids=('false-password', 'integer-password', 'too-long-password'),
```

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -178,6 +193,32 @@ def tls_certificate_private_key_pem_path(tls_certificate):
+        ),
+    )
+
+    with tls_certificate.private_key_pem.tempfile() as tf:
```
Changed to use tmp_path, but I am not happy with it, it leaves the files on my machine. The earlier versions cleaned the files nicely.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +798,124 @@ def test_http_over_https_error(
+        ('pyopenssl', 'tls_certificate_private_key_pem_path', False),
+    ),
+)
+# pylint: disable-next=R0917
```
Changed to human readable.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +798,124 @@ def test_http_over_https_error(
+        ('pyopenssl', 'tls_certificate_passwd_private_key_pem_path', True),
+        ('pyopenssl', 'tls_certificate_passwd_private_key_pem_path', False),
+        ('pyopenssl', 'tls_certificate_private_key_pem_path', True),
+        ('pyopenssl', 'tls_certificate_private_key_pem_path', False),
```
Added test ids, also changed the parameterization values.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +799,142 @@ def test_http_over_https_error(
+        'pyopenssl-non-encrypted-key-with-str-password',
+        'pyopenssl-non-encrypted-key-with-bytes-password',
+    ),
+)
```
This can be simplified to let pytest multiply the factors internally. Plus bytes/str IDs were swapped.
```suggestion
@pytest.mark.parametrize('adapter_type', ('builtin', 'pyopenssl'))
@pytest.mark.parametrize(
    'encrypted_key',
    (True, False),
    ids=('encrypted-key', 'unencrypted-key'),
)
@pytest.mark.parametrize(
    'password_as_bytes',
    (True, False),
    ids=('with-bytes-password', 'with-str-password'),
)
```

jatalahd replied:
Thank you, this was very helpful. Done in latest push.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +799,142 @@ def test_http_over_https_error(
+            ),
+        ),
+    ),
+    ids=('false-password', 'integer-password', 'too-long-password'),
```
I think "incorrect" is easier to put in the context than "false". At least, it sounds more obvious to me.
```suggestion
    ids=('incorrect-password', 'integer-password', 'too-long-password'),
```

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +799,142 @@ def test_http_over_https_error(
+):
+    """Check that builtin ssl-adapter initialization fails when wrong private key password given."""
+    tls_adapter_cls = get_ssl_adapter_class(name=adapter_type)
+    with pytest.raises(ssl.SSLError, match=r'\[SSL\] PEM.+'):
```
Is this the beginning of the string?
```suggestion
    with pytest.raises(ssl.SSLError, match=r'^\[SSL\] PEM.+'):
```

jatalahd replied:
Done in latest push.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -178,6 +192,34 @@ def tls_certificate_private_key_pem_path(tls_certificate):
+        ),
+    )
+
+    key_file = tmp_path / 'key.pem'
```
Done in latest push.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +799,142 @@ def test_http_over_https_error(
+            '@' * 2048,
+            pytest.warns(
+                UserWarning,
+                match=r'User-provided password is 2048 bytes.+',
```
Done in latest push.

jatalahd commented:
@webknjaz Provided improvements as you suggested. I hope this is OK now.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +799,141 @@ def test_http_over_https_error(
+    httpserver.stop()
+
+    interface, _host, port = _get_conn_data(
+        tls_http_server(httpserver.bind_addr, tls_adapter).bind_addr,
```
Why aren't you using the same server that was used to check TLS adapter loading the certs? This feels like having two different tests under the same roof. My idea was that the same server that loaded the certificates would be used for HTTPS requests and this working would be the final assertion that all's good.

webknjaz commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +799,141 @@ def test_http_over_https_error(
+    httpserver.stop()
+
+    interface, _host, port = _get_conn_data(
+        tls_http_server(httpserver.bind_addr, tls_adapter).bind_addr,
```
It seems this should be enough:
```suggestion
    key_file = (
        tls_certificate_passwd_private_key_pem_path
        if encrypted_key
        else tls_certificate_private_key_pem_path
    )
    key_pass = (
        private_key_password.encode('utf-8')
        if password_as_bytes
        else private_key_password
    )

    tls_adapter_cls = get_ssl_adapter_class(name=adapter_type)
    tls_adapter = tls_adapter_cls(
        certificate=tls_certificate_chain_pem_path,
        private_key=key_file,
        private_key_password=key_pass,
    )

    httpserver = tls_http_server(httpserver.bind_addr, tls_adapter)

    interface, _host, port = _get_conn_data(
        httpserver.bind_addr,
```

jatalahd replied:
Done as you suggested, but with a slight change that calling only tls_http_server inside the _get_conn_data().

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -757,3 +799,141 @@ def test_http_over_https_error(
+    httpserver = HTTPServer(
+        bind_addr=(ANY_INTERFACE_IPV4, EPHEMERAL_PORT),
+        gateway=Gateway,
+    )
```
Refactored in the latest push, now using just one call to tls_http_server fixture.

jatalahd commented:
@webknjaz Yet another improved push provided.

jatalahd commented on cheroot/test/test_ssl.py:
```diff
@@ -178,6 +192,34 @@ def tls_certificate_private_key_pem_path(tls_certificate):
+    private_key_password,
+    tmp_path,
+):
+    """Provide a certificate private key PEM file path via fixture."""
```
Used the suggested docstring in latest push.

webknjaz reviewed (approved):
Perfect, thanks!

webknjaz commented:
@jatalahd I've released this under v11.1.0:

* https://github.com/cherrypy/cheroot/actions/runs/19020536310
* https://github.com/cherrypy/cheroot/discussions/789
* https://github.com/cherrypy/cheroot/releases/tag/v11.1.0
* https://pypi.org/project/cheroot/11.1.0/

---

Though, one test got flaky and needs a follow-up: https://github.com/cherrypy/cheroot/pull/752/files#r2485127496.
Could you handle that, please?

---

Now, this allows sending a PR in the CherryPy repo per https://github.com/cherrypy/cherrypy/issues/1583#issuecomment-287476777.
